// @sap/cds/lib/utils/tests.js
const { resolve, dirname } = require('path')
const console = global.console

class Test extends require('./axios') {

  static run(..._) { return (new Test).run(..._) }
  get cds() { return require('../index') }
  get Test() { return Test }

  /**
   * Launches a cds server with arbitrary port and returns a subclass which
   * also acts as an axios lookalike, providing methods to send requests.
   */
  run (cmd, ...args) {

    // launch cds server...
    global.before (`launching ${cmd} ${args.join(' ')}...`, () => { // NOSONAR

      const {cds} = this

      if (!/^(serve|run)$/.test(cmd)) try {
        const project = cds.utils.isdir (cmd) || dirname (require.resolve (cmd+'/package.json'))
        cmd='serve'; args.push ('--project', project, '--in-memory?')
      } catch(e) {
        throw cds.error (`No such folder or package '${process.cwd()}' -> '${cmd}'`)
      }

      cds.once ('listening', ({server,url}) => {
        const axp = Reflect.getOwnPropertyDescriptor(this,'axios')
        if (axp) axp.value.defaults.baseURL = url
        this.server = server
        this.url = url
      })

      try { return cds.exec (cmd, ...args, '--port','0') }
      catch (e) { if (is_mocha) console.error(e) }
    })

    // shutdown cds server...
    global.after (done => {
      this.server ? this.server.close (done) : done()
    })

    return this
  }

  /**
   * Serving projects from subfolders under the root specified by a sequence
   * of path components which are concatenated with path.resolve().
   */
  in (...paths) {
    const {cds} = this; cds.root = resolve (cds.root, ...paths)
    // Checking conflicts with global.cds loaded in other folder before
    const ep = Reflect.getOwnPropertyDescriptor(global.cds,'env')
    if (ep && ep.value && ep.value._home !== cds.root) throw new Error (`[cds.test] - 'cds.env' was invoked before 'cds.test.in' from a different home:

      cds.env._home: ${cds.env._home}
      cds.test.in:   ${cds.root}

    > throwing this as tests would likely behave erratically.
    `)
    return this
  }

  /**
   * Switch on/off console log output.
   */
  verbose(v) {
    process.env.CDS_TEST_VERBOSE = v
    return this
  }

  /**
   * Lazily loads and returns an instance of chai
   */
  get chai() {
    const require = (mod) => { try { return module.require(mod) } catch(e) {
      if (e.code === 'MODULE_NOT_FOUND') throw new Error (`
      Failed to load required package '${mod}'. Please add it thru:
      npm add -D chai chai-as-promised chai-subset
    `)}}
    const chai = require('chai')
    chai.use (require('chai-subset'))
    chai.use (require('chai-as-promised'))
    return super.chai = chai
  }
  get expect(){ return this.chai.expect }
  get assert(){ return this.chai.assert }
  get sleep(){ return require('util').promisify(setTimeout) }
}


// harmonizing jest and mocha
const is_mocha = !!global.before
if (is_mocha) {
  const { format } = require('util')
  global.beforeAll = global.before
  global.afterAll = global.after
  global.test = global.it
  global.it.each = (table) => (title,fn) => Promise.all (table.map(each => {
      if (!Array.isArray(each)) each = [each]
      return it (format(title,...each), ()=> fn(...each))
  }))
} else { // it's jest
  global.before = (msg,fn) => global.beforeAll(fn||msg)
  global.after = (msg,fn) => global.afterAll(fn||msg)
}

if (!process.env.CDS_TEST_VERBOSE)  silent()

function silent() {
  const console = global.console, logs=[]
  global.console = { __proto__: global.console, logs,
    time: ()=>{}, timeEnd: (...args)=> logs.push(args),
    debug: (...args)=> logs.push(args),
    log: (...args)=> logs.push(args),
    warn: (...args)=> logs.push(args),
    error: (...args)=> logs.push(args),
    dump(){ for (let each of logs) console.log (...each) },
  }
  global.after (()=> global.console = console)
}


/** @type Test.run & Test */
module.exports = Object.setPrototypeOf (Test.run, Test.prototype)
